fireant/queries/builder.py

```diff
@@ -297,16 +301,49 @@ def sql(self):
         Serializes this query builder as a set of SQL queries.  This method will always return a list of one query since
         only one query is required to retrieve dimension choices.
 
-        This function only handles dimensions (select+group by) and filtering (where/having), which is everything needed
-        for the query to fetch choices for dimensions.
-
         The slicer query extends this with metrics, references, and totals.
         """
+
+        def in_hint_table(dim):
+            return all(dim_field.name in hint_column_names for dim_field in dim.definition.fields())
+
+        hint_column_definitions = self.dataset.database.get_column_definitions(
+              self.dataset.hint_table._schema._name,
+              self.dataset.hint_table._table_name
+        ) if self.dataset.hint_table else []
+
+        hint_column_names = [hint_column_definition[0] for hint_column_definition in hint_column_definitions]
+
+        dimensions = self._dimensions
+
+        # All definitions of all dimensions have to exist in hint table
+        is_hint_dimension = all(in_hint_table(dim) for dim in self._dimensions)
+
+        table = self.dataset.hint_table if is_hint_dimension else self.dataset.table
+        joins = () if is_hint_dimension else self.dataset.joins
+        filters = self._filters
+
+        if is_hint_dimension:
+            # Prevent definition changes to the original dimensions
+            dimensions = [copy.deepcopy(dimension) for dimension in self._dimensions]
+
+            for dimension in dimensions:
+                for dimension_field in dimension.definition.fields():
+                    dimension_field.table = dimension_field.for_(table).table
+
+            filters = [hint_filter
+                       for hint_filter in filters
+                       if in_hint_table(hint_filter)]
+
+            for hint_filter in filters:
```

Comment:
it might be easier to just do this for the whole query instead of for each field